docs/service-bundle.md

```diff
@@ -0,0 +1,225 @@
+* __POD_NAME__: the name of the pod 
+
+## Output
+
+### stdout / stderr
+
+These are ignored.
+
+### Exit Code
+
+* 0 - success
+* 1 - error
+* 8 - action not implemented/supported
+
+### Encoded Credentials
+
+Binding credentials may be returned during a provision or bind operation by
+creating a named secret (POD_NAME) in the sandbox namespace (POD_NAMESPACE).
+The secret should have a single key called “fields” whose value is a
+base64-encoded JSON document.
+
+Ansible Playbook Bundles use the “asb_encode_binding” module to do this.
+
+```
+apiVersion: v1
+data:
+  fields: eyJEQl9OQU1FIjogImZvbyIsICJEQl9QQVNTV09SRCI6ICJsZXRtZWluIiwgIkRCX1RZUEUiOiAicG9zdGdyZXMiLCAiREJfUE9SVCI6ICI1NDMyIiwgIkRCX1VTRVIiOiAiYmFyIiwgIkRCX0hPU1QiOiAicG9zdGdyZXNxbCJ9
+kind: Secret
+metadata:
+  creationTimestamp: 2018-02-26T18:34:50Z
+  name: apb-874ae78d-acfe-4181-b064-124d6e13c592
+  namespace: dh-postgresql-apb-prov-md5xt
+  resourceVersion: "11846"
+  selfLink: /api/v1/namespaces/dh-postgresql-apb-prov-md5xt/secrets/apb-874ae78d-acfe-4181-b064-124d6e13c592
+  uid: ba79e27a-1b23-11e8-aebf-94b96c9ea8ca
+type: Opaque
+```
+
+The value for “fields” is a base64-encoded JSON document whose keys and values
+represent any data the APB wants to return to the broker and ultimately to a
+client that retrieves a binding.
+
+__Provision Credentials__ - If a service bundle saves such a secret during the
+provision operation, the broker saves the passed data as “provision
+credentials”. That data gets passed into all further actions as part of the
+JSON document under the key ``_apb_provision_creds``, as a sub-document. As such,
+it can be useful for making root-level credentials created during provision
+available to other operations that need to use them. If a request for non-async
+binding is made and the bundle does not run, any “provision credentials” that
+were saved will be returned to the client as binding credentials.
+
+__Binding Credentials__ - If a service bundle saves such a secret during the bind
+operation, the broker saves the passed data as “binding credentials”. These
+credentials get returned to the client that initiated that specific binding
+operation.
+
+## Runtime
+
+The container will be run in a pod by itself with ``restartPolicy: Never``. The
+pod will run in a sandbox namespace created by the broker for that specific
+operation only. The namespace will be deleted at some point after the pod
+finishes.
+
+APB is run with a service account granted access to the ``POD_NAMESPACE`` and
+to the “namespace” that the pod is to deploy to. Networks between the two
+namespaces are tied together so that you can reliably find your services in the
+target namespace. That only works for
+[NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
+or for multi tenant openshift-sdn.
+
+### Proxy
+
+If there are proxy settings for the broker, they will be forwarded to the pod
```

Comment:
This is correct, but doesn't entirely address the story. We also include lowercased variants, which is actually technically important because there are a lot of ansible modules (more specifically, python libraries) that are utterly inconsistent in their expectations, or downright don't respect the presence of these env vars. The first boto being one of them.

In addition, we take no action to pass on these variables to the app pods that are spawned by your APB. That means you are responsible for writing your APB to check for proxy settings in their environment, and conditionally applying them to your service if it is required. There is interest in some kind of framework support for delivering these settings automatically, but that's not currently available.

Reply:
Wow. It is good to know that. That reminds me that it may also be a requirement of a Service Bundle developer to annotate the objects they create in the cluster a particular way, like the service instance they belong to.

Reply:
If we are moving to generic definition of service bundle then perhaps its not even worth mentioning in the contract document. The contract should be agnostic of what the bundle is actually doing. In this case we would be assuming a service bundle is deploying on-cluster services.

Perhaps we also create a "Best Practices" document for provision on-cluster services? Unsure.

Reply:
The hope is that we might be able to use PodPresets to deliver them automatically via label, and then we could write a module that wraps that to make it easy for developers.